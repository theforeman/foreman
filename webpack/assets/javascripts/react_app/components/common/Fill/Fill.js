import React from 'react';
import PropTypes from 'prop-types';

class Fill extends React.Component {
  componentDidMount() {
    const {
      children,
      overrideProps,
      registerFillComponent,
      slotId,
      weight,
      id,
    } = this.props;

    registerFillComponent(slotId, overrideProps, id, children, weight);
  }
  componentWillUnmount() {
    const { slotId, unregisterFillComponent, id } = this.props;

    unregisterFillComponent(slotId, id);
  }
  render() {
    return null;
  }
}

Fill.propTypes = {
  // a component to be injected on a slot
  children: PropTypes.oneOfType([PropTypes.node, PropTypes.object]),
  registerFillComponent: PropTypes.func.isRequired,
  unregisterFillComponent: PropTypes.func.isRequired,
  slotId: PropTypes.string.isRequired,
  // ordering between slot's fills, higher will be rendered first
  weight: PropTypes.number.isRequired,
  // fill's id
  id: PropTypes.string.isRequired,
  // a props object to be injected on the slot's children
  overrideProps: PropTypes.object,
};

Fill.defaultProps = {
  children: undefined,
  overrideProps: undefined,
};

export default Fill;
