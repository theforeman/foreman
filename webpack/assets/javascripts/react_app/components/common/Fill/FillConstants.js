export const REGISTER_FILL = 'SLOT_AND_FILL_REGISTER_FILL';

export const REMOVE_FILLED_COMPONENT = 'SLOT_AND_FILL_REMOVE_FILLED_COMPONENT';
