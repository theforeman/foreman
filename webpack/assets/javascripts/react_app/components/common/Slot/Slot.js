import { cloneElement, isValidElement } from 'react';
import PropTypes from 'prop-types';

const Slot = ({ fills, id, multi, children = null, ...props }) => {
  const addProps = object => {
    if (multi && !object.key) {
      // eslint-disable-next-line no-console
      console.warn('Please add a key attribute to multiple fills');
    }

    if (isValidElement(object)) {
      return cloneElement(object, { ...props });
    }

    if (!children) {
      throw new Error('Slot with override props must have a child');
    }

    return cloneElement(children, { ...props, ...object });
  };

  if (fills.length) return fills.map(component => addProps(component));
  return children;
};

Slot.propTypes = {
  fills: PropTypes.array,
  id: PropTypes.string.isRequired,
  multi: PropTypes.bool,
  children: PropTypes.node,
};

Slot.defaultProps = {
  fills: [],
  multi: false,
  children: undefined,
};

export default Slot;
