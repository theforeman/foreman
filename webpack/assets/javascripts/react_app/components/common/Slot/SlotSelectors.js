import { getSlotComponents } from '../../../../services/SlotsRegistry';

export const selectComponentByWeight = slotId =>
  getSlotComponents(slotId)
    .sort((a, b) => b.weight - a.weight)
    .map(c => c.component);

export const selectMaxComponent = slotId => selectComponentByWeight(slotId)[0];

export const selectFillsAmount = (state, id) => {
  const registerdFills = state.extendable[id];
  return registerdFills ? Object.keys(registerdFills).length : 0;
};

export const selectFillsComponents = (state, props) => {
  const { id, multiple } = props;

  if (selectFillsAmount(state, id)) {
    if (multiple) return selectComponentByWeight(id);
    return [selectMaxComponent(id)];
  }
  return [];
};
