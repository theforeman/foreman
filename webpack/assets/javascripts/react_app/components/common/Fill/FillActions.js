import { REGISTER_FILL, REMOVE_FILLED_COMPONENT } from './FillConstants';
import { add, remove } from '../../../../services/SlotsRegistry';

export const registerFillComponent = (
  slotId,
  overrideProps,
  fillId,
  component,
  weight
) => dispatch => {
  add(slotId, fillId, component, weight, overrideProps);
  dispatch({
    type: REGISTER_FILL,
    payload: { slotId, fillId },
  });
};

export const unregisterFillComponent = (slotId, fillId) => dispatch => {
  remove(slotId, fillId);
  dispatch({
    type: REMOVE_FILLED_COMPONENT,
    payload: { slotId },
  });
};
