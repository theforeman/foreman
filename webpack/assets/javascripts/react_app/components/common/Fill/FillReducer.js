import Immutable from 'seamless-immutable';

import { REGISTER_FILL, REMOVE_FILLED_COMPONENT } from './FillConstants';

const initialState = Immutable({});

export default (state = initialState, action) => {
  const { payload } = action;

  switch (action.type) {
    case REGISTER_FILL:
      return state.setIn([payload.slotId, payload.fillId], true);

    case REMOVE_FILLED_COMPONENT:
      return state.setIn([payload.slotId, payload.fillId], false);
    default:
      return state;
  }
};
