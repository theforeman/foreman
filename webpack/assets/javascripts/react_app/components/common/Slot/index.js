import { connect } from 'react-redux';
import { selectFillsComponents } from './SlotSelectors';
import Slot from './Slot';

// map state to props
const mapStateToProps = (state, ownProps) => ({
  fills: selectFillsComponents(state, {
    id: ownProps.id,
    multiple: ownProps.multi,
  }),
});

// export connected component
export default connect(mapStateToProps)(Slot);
