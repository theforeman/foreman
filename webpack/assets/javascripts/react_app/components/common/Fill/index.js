import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import * as actions from './FillActions';
import reducer from './FillReducer';

import Fill from './Fill';

// map action dispatchers to props
const mapDispatchToProps = dispatch => bindActionCreators(actions, dispatch);

// export reducers
export const reducers = { extendable: reducer };

// export connected component
export default connect(null, mapDispatchToProps)(Fill);
