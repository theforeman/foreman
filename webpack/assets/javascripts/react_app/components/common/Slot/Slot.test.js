import React from 'react';
import IntegrationTestHelper from '../../../common/IntegrationTestHelper';

import Fill from '../Fill';
import Slot from './';
import FillReducer from '../Fill/FillReducer';

const combinedReducers = { extendable: FillReducer };
const SlotComponent = ({ text }) => <div>{text}</div>; // eslint-disable-line

jest.unmock('../../../../services/SlotsRegistry');

describe('Slot-Fill', () => {
  const integrationTestHelper = new IntegrationTestHelper(combinedReducers);

  it('render multiple fills', () => {
    const FirstComponent = () => <span> Should be the first in the snap</span>;
    const SecondComponent = () => <span> Should be the first in the snap</span>;
    const wrapper = integrationTestHelper.mount(
      <React.Fragment>
        <Fill slotId="slot-1" id="some-key-1" weight={100}>
          <SecondComponent key={1} />
        </Fill>
        <Fill slotId="slot-1" id="some-key-2" weight={200}>
          <FirstComponent key={2} />
        </Fill>

        <Slot id="slot-1" multi>
          <div> defaul value </div>
        </Slot>
      </React.Fragment>
    );

    wrapper.update();
    expect(wrapper.find('Slot')).toMatchSnapshot();
    integrationTestHelper.takeStoreSnapshot();
    integrationTestHelper.takeActionsSnapshot();
  });

  it('Slot with a prop should be pass to a fill', () => {
    const wrapper = integrationTestHelper.mount(
      <React.Fragment>
        <Fill slotId="slot-9" id="some-key-1" weight={100}>
          <SlotComponent key="key" />
        </Fill>

        <Slot id="slot-9" text="this prop should be taken" />
      </React.Fragment>
    );

    wrapper.update();
    expect(wrapper.find('Slot')).toMatchSnapshot();
  });

  it('Fill with no component nor overriden props should throw an error', () => {
    // eslint-disable-next-line no-console
    const err = console.error;
    // eslint-disable-next-line no-console
    console.error = jest.fn();
    expect(() => {
      integrationTestHelper.mount(
        <React.Fragment>
          <Fill slotId="slot-7" id="some-key-1" weight={100} />

          <Slot id="slot-7" />
        </React.Fragment>
      );
    }).toThrowError(new Error('Slot with override props must have a child'));
    // eslint-disable-next-line no-console
    console.error = err;
  });

  it('no multiple fills', () => {
    const AbsentComponent = () => <div> This should not be in the snap </div>;
    const PresentComponent = () => (
      <span> This span should be in the snap </span>
    );
    const wrapper = integrationTestHelper.mount(
      <React.Fragment>
        <Fill slotId="slot-2" id="some-key-1" weight={100}>
          <AbsentComponent key="a" />
        </Fill>
        <Fill slotId="slot-2" id="some-key-1" weight={200}>
          <PresentComponent key="b" />
        </Fill>

        <Slot id="slot-2" multi={false}>
          <div> defaul value </div>
        </Slot>
      </React.Fragment>
    );

    wrapper.update();
    expect(wrapper.find('Slot')).toMatchSnapshot();
  });
  it('props fill', () => {
    const wrapper = integrationTestHelper.mount(
      <React.Fragment>
        <Fill
          overrideProps={{ text: 'This is given by a prop' }}
          slotId="slot-3"
          id="some-key-1"
          weight={100}
        />

        <Slot id="slot-3" multi={false}>
          <SlotComponent key="c" />
        </Slot>
      </React.Fragment>
    );

    wrapper.update();
    expect(wrapper.find('Slot')).toMatchSnapshot();
  });
  it('default slot', () => {
    const wrapper = integrationTestHelper.mount(
      <Slot id="slot-4" multi>
        <SlotComponent text="Default Value" />
      </Slot>
    );

    wrapper.update();
    expect(wrapper.find('Slot')).toMatchSnapshot();
  });
  it('multiple slot with override props', () => {
    const wrapper = integrationTestHelper.mount(
      <React.Fragment>
        <Fill
          overrideProps={{ text: 'This should be the second', key: 1 }}
          slotId="slot-5"
          id="some-key-1"
          weight={100}
        />
        <Fill
          overrideProps={{ text: 'This should be the first', key: 2 }}
          slotId="slot-5"
          id="some-key-2"
          weight={200}
        />
        <Slot id="slot-5" multi>
          <SlotComponent text="Default Value" />
        </Slot>
      </React.Fragment>
    );
    wrapper.update();
    expect(wrapper.find('Slot')).toMatchSnapshot();
  });

  it('slot with multi override props should take max weight', () => {
    const wrapper = integrationTestHelper.mount(
      <React.Fragment>
        <Fill
          overrideProps={{ text: 'This should not be in the snap' }}
          slotId="slot-6"
          id="some-key-1"
          weight={100}
        />
        <Fill
          overrideProps={{
            text: 'This text should be in the snap',
            key: 'textKey',
          }}
          slotId="slot-6"
          id="some-key-2"
          weight={200}
        />
        <Slot id="slot-6">
          <SlotComponent text="Default Value" />
        </Slot>
      </React.Fragment>
    );
    wrapper.update();
    expect(wrapper.find('Slot')).toMatchSnapshot();
  });

  it('multi slot with override props fill and component fill', () => {
    const TestComponent = () => <div> Also this should be in the snap </div>;
    const wrapper = integrationTestHelper.mount(
      <React.Fragment>
        <Fill
          overrideProps={{ key: 'def', text: 'This should be in the snap' }}
          slotId="slot-10"
          id="some-key-1"
          weight={100}
        />
        <Fill slotId="slot-10" id="some-key-2" weight={100}>
          <TestComponent key="abc" />
        </Fill>
        <Slot id="slot-10" multi>
          <SlotComponent text="Default Value" />
        </Slot>
      </React.Fragment>
    );
    wrapper.update();
    expect(wrapper.find('Slot')).toMatchSnapshot();
  });
});
