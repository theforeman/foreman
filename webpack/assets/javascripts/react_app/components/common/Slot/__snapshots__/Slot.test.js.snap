// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slot-Fill Slot with a prop should be pass to a fill 1`] = `
<Slot
  dispatch={[Function]}
  fills={
    Array [
      <SlotComponent />,
    ]
  }
  id="slot-9"
  multi={false}
  text="this prop should be taken"
>
  <SlotComponent
    dispatch={[Function]}
    key="key"
    text="this prop should be taken"
  >
    <div>
      this prop should be taken
    </div>
  </SlotComponent>
</Slot>
`;

exports[`Slot-Fill default slot 1`] = `
<Slot
  dispatch={[Function]}
  fills={Array []}
  id="slot-4"
  multi={true}
>
  <SlotComponent
    text="Default Value"
  >
    <div>
      Default Value
    </div>
  </SlotComponent>
</Slot>
`;

exports[`Slot-Fill multi slot with override props fill and component fill 1`] = `
<Slot
  dispatch={[Function]}
  fills={
    Array [
      Object {
        "key": "def",
        "text": "This should be in the snap",
      },
      <TestComponent />,
    ]
  }
  id="slot-10"
  multi={true}
>
  <SlotComponent
    dispatch={[Function]}
    key="def"
    text="This should be in the snap"
  >
    <div>
      This should be in the snap
    </div>
  </SlotComponent>
  <TestComponent
    dispatch={[Function]}
    key="abc"
  >
    <div>
       Also this should be in the snap 
    </div>
  </TestComponent>
</Slot>
`;

exports[`Slot-Fill multiple slot with override props 1`] = `
<Slot
  dispatch={[Function]}
  fills={
    Array [
      Object {
        "key": 2,
        "text": "This should be the first",
      },
      Object {
        "key": 1,
        "text": "This should be the second",
      },
    ]
  }
  id="slot-5"
  multi={true}
>
  <SlotComponent
    dispatch={[Function]}
    key="2"
    text="This should be the first"
  >
    <div>
      This should be the first
    </div>
  </SlotComponent>
  <SlotComponent
    dispatch={[Function]}
    key="1"
    text="This should be the second"
  >
    <div>
      This should be the second
    </div>
  </SlotComponent>
</Slot>
`;

exports[`Slot-Fill no multiple fills 1`] = `
<Slot
  dispatch={[Function]}
  fills={
    Array [
      <PresentComponent />,
    ]
  }
  id="slot-2"
  multi={false}
>
  <PresentComponent
    dispatch={[Function]}
    key="b"
  >
    <span>
       This span should be in the snap 
    </span>
  </PresentComponent>
</Slot>
`;

exports[`Slot-Fill props fill 1`] = `
<Slot
  dispatch={[Function]}
  fills={
    Array [
      Object {
        "text": "This is given by a prop",
      },
    ]
  }
  id="slot-3"
  multi={false}
>
  <SlotComponent
    dispatch={[Function]}
    key="c"
    text="This is given by a prop"
  >
    <div>
      This is given by a prop
    </div>
  </SlotComponent>
</Slot>
`;

exports[`Slot-Fill render multiple fills 1`] = `
<Slot
  dispatch={[Function]}
  fills={
    Array [
      <FirstComponent />,
      <SecondComponent />,
    ]
  }
  id="slot-1"
  multi={true}
>
  <FirstComponent
    dispatch={[Function]}
    key="2"
  >
    <span>
       Should be the first in the snap
    </span>
  </FirstComponent>
  <SecondComponent
    dispatch={[Function]}
    key="1"
  >
    <span>
       Should be the first in the snap
    </span>
  </SecondComponent>
</Slot>
`;

exports[`Slot-Fill render multiple fills 2`] = `
Object {
  "extendable": Object {
    "slot-1": Object {
      "some-key-1": true,
      "some-key-2": true,
    },
  },
}
`;

exports[`Slot-Fill render multiple fills: Integration test actions 1`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "fillId": "some-key-1",
        "slotId": "slot-1",
      },
      "type": "SLOT_AND_FILL_REGISTER_FILL",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "fillId": "some-key-2",
        "slotId": "slot-1",
      },
      "type": "SLOT_AND_FILL_REGISTER_FILL",
    },
  ],
]
`;

exports[`Slot-Fill slot with multi override props should take max weight 1`] = `
<Slot
  dispatch={[Function]}
  fills={
    Array [
      Object {
        "key": "textKey",
        "text": "This text should be in the snap",
      },
    ]
  }
  id="slot-6"
  multi={false}
>
  <SlotComponent
    dispatch={[Function]}
    key="textKey"
    text="This text should be in the snap"
  >
    <div>
      This text should be in the snap
    </div>
  </SlotComponent>
</Slot>
`;
